File: backend/app/Domain/Calculators/Services/AreaUnitConverter.php
<?php

namespace App\Domain\Calculators\Services;

use InvalidArgumentException;

/**
 * Single source of truth for area unit conversions. Nepal listings commonly use
 * ropani/aana in the hills (Kathmandu Valley, Pokhara) and kattha/dhur in the
 * Terai (Chitwan, Biratnagar) alongside sqft/sqm — every screen that shows or
 * accepts an area should convert through here rather than hardcoding factors.
 *
 * Conversion factors are Nepal's official survey department constants.
 */
class AreaUnitConverter
{
    private const TO_SQM = [
        'sqm' => 1.0,
        'sqft' => 0.092903,
        'aana' => 31.7949,
        'ropani' => 508.72,
        'kattha' => 338.63,
        'dhur' => 16.93,
    ];

    public static function toSqm(float $value, string $unit): float
    {
        return round($value * self::factor($unit), 2);
    }

    public static function fromSqm(float $sqm, string $unit): float
    {
        return round($sqm / self::factor($unit), 2);
    }

    private static function factor(string $unit): float
    {
        if (! isset(self::TO_SQM[$unit])) {
            throw new InvalidArgumentException("Unknown area unit [{$unit}].");
        }

        return self::TO_SQM[$unit];
    }

    public static function units(): array
    {
        return array_keys(self::TO_SQM);
    }
}

File: mobile/lib/core/formatters/area_unit_converter.dart
/// Nepal land-area unit conversions, ported exactly from the backend's
/// single source of truth:
/// backend/app/Domain/Calculators/Services/AreaUnitConverter.php
///
/// Nepal listings commonly use ropani/aana in the hills (Kathmandu Valley,
/// Pokhara) and kattha/dhur in the Terai (Chitwan, Biratnagar) alongside
/// sqft/sqm. Conversion factors are Nepal's official survey department
/// constants — every screen showing or accepting an area should convert
/// through here rather than hardcoding factors.
class AreaUnitConverter {
  AreaUnitConverter._();

  static const Map<String, double> _toSqm = {
    'sqm': 1.0,
    'sqft': 0.092903,
    'aana': 31.7949,
    'ropani': 508.72,
    'kattha': 338.63,
    'dhur': 16.93,
  };

  static double toSqm(double value, String unit) {
    return _round2(value * _factor(unit));
  }

  static double fromSqm(double sqm, String unit) {
    return _round2(sqm / _factor(unit));
  }

  static double _factor(String unit) {
    final factor = _toSqm[unit];
    if (factor == null) {
      throw ArgumentError('Unknown area unit [$unit].');
    }
    return factor;
  }

  static double _round2(double value) => (value * 100).round() / 100;

  static List<String> get units => _toSqm.keys.toList(growable: false);
}
